ark/phenotyping/som_utils.py

```diff
@@ -0,0 +1,131 @@
+    misc_utils.verify_in_list(fovs=fovs, image_fovs=img_xr.fovs.values)
+    misc_utils.verify_in_list(channels=channels, image_channels=img_xr.channels.values)
+
+    # define our flowsom matrix
+    flowsom_data = None
+
+    # iterate over fovs
+    for fov in fovs:
+        # subset img_xr with only the fov we're looking for, and cast to float32
+        img_data_blur = img_xr.loc[fov, ..., channels].values.astype(np.float32)
+
+        # for each marker, compute the Gaussian blur
+        for marker in range(len(channels)):
+            img_data_blur[:, :, marker] = ndimage.gaussian_filter(img_data_blur[:, :, marker],
+                                                                  sigma=blur_factor)
+
+        # flatten each image
+        pixel_mat = img_data_blur.reshape(-1, len(channels))
+
+        # convert into a dataframe
+        pixel_mat = pd.DataFrame(pixel_mat, columns=channels)
+
+        # assign metadata about each entry
+        pixel_mat['fov'] = fov
+        pixel_mat['row_index'] = np.repeat(range(img_data_blur.shape[0]), img_data_blur.shape[1])
+        pixel_mat['column_index'] = np.tile(range(img_data_blur.shape[0]), img_data_blur.shape[1])
+
+        # assign segmentation label
+        seg_labels_flat = seg_labels.loc[fov, ...].values.flatten()
+        pixel_mat['segmentation_label'] = seg_labels_flat
+
+        # remove any rows that sum to 0
+        pixel_mat = pixel_mat.loc[pixel_mat.loc[:, channels].sum(axis=1) != 0, :]
+
+        # normalize each row by total marker counts to convert into frequencies
+        pixel_mat.loc[:, channels] = pixel_mat.loc[:, channels].div(
+            pixel_mat.loc[:, channels].sum(axis=1), axis=0)
+
+        # assign to flowsom_data if not already assigned, otherwise concatenates
+        if flowsom_data is None:
+            flowsom_data = pixel_mat
+        else:
+            flowsom_data = pd.concat([flowsom_data, pixel_mat])
+
+    # normalize each marker column by the 99.9 percentile value
+    flowsom_data.loc[:, channels] = flowsom_data.loc[:, channels].div(
+        flowsom_data.loc[:, channels].quantile(q=0.999, axis=0), axis=1)
+
+    return flowsom_data
+
+
+def cluster_pixels(chan_list, base_dir, pixel_pre_name=None, pixel_cluster_name=None):
+    """Run the FlowSOM training on the pixel data.
+
+    Saves results to pixel_mat_clustered.csv in base_dir.
+    Usage: Rscript som_runner.R {path_to_pixel_matrix} {chan_list_comma_separated} {save_path}
+
+    Args:
+        chan_list (list):
+            The list of markers to subset on
+        base_dir (str):
+            The path to the directory to save the clustered pixel matrix in
+        pixel_pre_name (str):
+            The name of the preprocessed file name
+        pixel_cluster_name (str):
+            The name of the file to write the clustered csv to
+    """
+
+    # set the path to the preprocessed matrix, default to pixel_mat_preprocessed.csv
+    if pixel_pre_name is None:
```

Comment:
Rather than setting the default to None and replacing if it's not None, we should just set the default to the value that you assign here